kyaml/yaml/fns_test.go

```diff
@@ -97,6 +97,76 @@ func TestGetElementByKey(t *testing.T) {
 	assert.Equal(t, "f: g\n", assertNoErrorString(t)(rn.String()))
 }
 
+func TestElementSetter(t *testing.T) {
+	orig := MustParse(`
+- a: b
+- scalarValue
+- c: d
+# null will be removed
+- null
+`)
+
+	// ElementSetter will update node, so make a copy
+	node := orig.Copy()
+	// Remove an element
+	rn, err := node.Pipe(ElementSetter{Key: "a", Value: "b"})
+	assert.NoError(t, err)
+	assert.Nil(t, rn)
+	assert.Equal(t, `- scalarValue
+- c: d
+`, assertNoErrorString(t)(node.String()))
+
+	node = orig.Copy()
+	// Set an element
+	newElement := NewMapRNode(&map[string]string{
+		"e": "f",
+	})
+	rn, err = node.Pipe(ElementSetter{
+		Key:     "a",
+		Value:   "b",
+		Element: newElement.YNode(),
+	})
+	assert.NoError(t, err)
+	assert.Equal(t, rn, newElement)
+	assert.Equal(t, `- e: f
+- scalarValue
+- c: d
+`, assertNoErrorString(t)(node.String()))
+
+	node = orig.Copy()
+	// Set an element with scalar
+	newElement = NewScalarRNode("foo")
+	rn, err = node.Pipe(ElementSetter{
+		Key:     "a",
+		Value:   "b",
+		Element: newElement.YNode(),
+	})
+	assert.NoError(t, err)
+	assert.Equal(t, rn, newElement)
+	assert.Equal(t, `- foo
+- scalarValue
+- c: d
+`, assertNoErrorString(t)(node.String()))
+
+	node = orig.Copy()
+	// Append an element
+	newElement = NewMapRNode(&map[string]string{
+		"e": "f",
+	})
+	rn, err = node.Pipe(ElementSetter{
+		Key:     "x",
+		Value:   "y",
+		Element: newElement.YNode(),
+	})
```

Comment:
This filter's behavior is too clever and non-obvious.

The 'x' and 'y' don't appear in the result at all here, which is not what one would expect without reading all the code.

ElementFilter is like a search and replace if found, else append if not found, and the definition of found is complicated since the key must match but maybe we don't match the value, and the behavior becomes a delete if the Element is left nil.

Suggestion:  the doc on the filter should enumerate some of these examples, e,g

```
ElementSetter will append, replace or delete an element in an associative list.
To append, ...
To replace, ...
To delete, ...
```